src/net_doh.c

```diff
@@ -0,0 +1,1140 @@
+    debugx("provider_read_cb - len: %d, stream_id: %d", length, stream_id);
+    return 0;
+}
+
+static int _http2_header_cb(nghttp2_session* session, const nghttp2_frame* frame, const uint8_t* name, size_t namelen, const uint8_t* value, size_t valuelen, uint8_t flags, void* user_data)
+{
+    perf__doh_socket_t *sock = (perf__doh_socket_t *)user_data;
+    (void)flags;
+
+    // // debugx("header_cb - type: %d, session_id: %d", frame->hd.type, frame->hd.stream_id);
+
+    switch (frame->hd.type) {
+    case NGHTTP2_HEADERS:
+        /* TODO: remove
+        if (frame->headers.cat == NGHTTP2_HCAT_RESPONSE &&
+            self->http2->stream->stream_id == frame->hd.stream_id) {
+            fwrite(name, 1, namelen, stderr);
+            fprintf(stderr, ": ");
+            fwrite(value, 1, valuelen, stderr);
+            fprintf(stderr, "\n");        
+        }
+        */
+       break;
+    }
+    return 0;
+}
+
+static int _http2_stream_close_cb(nghttp2_session* session, int32_t stream_id, uint32_t error_code, void* user_data)
+{
+    (void)user_data;
+
+    // TODO:
+    // // debugx("close_cb - stream_id: %d", stream_id);
+    
+    if (nghttp2_session_get_stream_user_data(session, stream_id)) {
+        // debugx("http2 session closed - stream_id: %d, error_code: %d", 
+        //        stream_id, error_code);
+        int ret;
+        ret = nghttp2_session_terminate_session(session, NGHTTP2_NO_ERROR);
+
+        if (ret != 0) {
+            return NGHTTP2_ERR_CALLBACK_FAILURE;
+        }
+    }
+
+    return 0;
+}
+
+static int _http2_frame_recv_cb(nghttp2_session* session, const nghttp2_frame* frame, void* user_data)
+{
+    perf__doh_socket_t *sock = (perf__doh_socket_t *)user_data;
+
+    // TODO:
+    // debugx("frame_recv_cb - type: %d, stream_id: %d", frame->hd.type, frame->hd.stream_id);
+    switch (frame->hd.type) {
+    case NGHTTP2_HEADERS:
+        /* TODO: remove
+        if (frame->headers.cat == NGHTTP2_HCAT_RESPONSE) {
+            const nghttp2_nv *nva = frame->headers.nva;
+            if (nghttp2_session_get_stream_user_data(session, frame->hd.stream_id)) {
+                for (size_t i = 0; i < frame->headers.nvlen; ++i) {
+                    fwrite(nva[i].name, 1, nva[i].namelen, stderr);
+                    fprintf(stderr, ": ");
+                    fwrite(nva[i].value, 1, nva[i].valuelen, stderr);
+                    fprintf(stderr, "\n");
+                }
+            }
+        }
+        */
+        break;
+    case NGHTTP2_DATA: 
+        // debugx("DATA frame received - flags: %d", frame->hd.flags);
+        // we are interested in DATA frame which will carry the DNS response
+        // NGHTTP2_FLAG_END_STREAM indicates that we have the data in full
+        if (frame->hd.flags & NGHTTP2_FLAG_END_STREAM) {
+            // debugx("END_STREAM - copying to recvbuf\n");
+            if (self->http2->dnsmsg_at > TCP_RECV_BUF_SIZE) {
+                perf_log_warning("DNS response > TCP_RECV_BUF_SIZE");
```

Comment:
you need to fix this, DNS message can be large then the buffer.

You can't just blindly copy the DoT code like this.

Reply:
Why are you checking against `TCP_RECV_BUF_SIZE` btw? `http2->dnsmsg` is size of `DNS_MSG_MAX_SIZE`

Reply:
I left TODO note on this in the code previously to review as I left it based on the recvbuf size in the sock struct which is not in play - will amend.